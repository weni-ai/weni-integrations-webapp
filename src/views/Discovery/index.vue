<template>
  <div class="discovery-content">
    <unnnic-input
      v-model="searchTerm"
      class="discovery-content__search"
      :placeholder="$t('apps.discovery.search.placeholder')"
      icon-left="search-1"
    />

    <span v-if="searchTerm && searchTerm.trim()" class="discovery-content__search__results">
      {{ $t('apps.discovery.search.results') }}
      <span class="discovery-content__search__results__highlight">
        {{ `“${searchTerm}”...` }}
      </span>
    </span>
    <div v-if="hasAnyVisibleApp" class="discovery-content__grids">
      <AppGrid
        ref="appGrid"
        section="channel"
        type="add"
        :loading="loadingAllAppTypes"
        :apps="filteredApps"
        @update="fetchChannels"
      />

      <AppGrid
        section="ecommerce"
        type="add"
        :loading="loadingEcommerceApps"
        :apps="filteredEcommerceApps"
        @update="fetchEcommerceApps"
      />

      <AppGrid
        section="external"
        type="add"
        :loading="loadingExternalServices"
        :apps="filteredExternalServices"
        @update="fetchExternalServices"
      />
    </div>
    <div v-else-if="searchTerm && !filteredApps.length && !filteredExternalServices.length">
      <EmptyApps :term="searchTerm" />
    </div>

    <div v-if="searchTerm" class="discovery-content__recommended">
      <AppGrid
        section="recommended"
        type="add"
        :loading="loadingFeaturedApps"
        :apps="featuredApps"
      />
    </div>
  </div>
</template>

<script>
  import AppGrid from '@/components/AppGrid/index.vue';
  import EmptyApps from '@/components/EmptyApps/index.vue';
  import { mapActions, mapState } from 'pinia';
  import { app_type } from '@/stores/modules/appType/appType.store';
  import { externals_store } from '@/stores/modules/appType/externals/externals.store';
  import { ecommerce_store } from '@/stores/modules/appType/ecommerce/ecommerce.store';
  import unnnic from '@weni/unnnic-system';
  export default {
    name: 'Discovery',
    components: {
      AppGrid,
      EmptyApps,
    },
    data() {
      return {
        searchTerm: '',
        channels: {
          loading: true,
          data: null,
        },
      };
    },
    async mounted() {
      this.fetchChannels();

      this.fetchExternalServices();

      this.fetchEcommerceApps();

      this.fetchFeatured();
    },
    computed: {
      ...mapState(app_type, [
        'allAppTypes',
        'loadingAllAppTypes',
        'errorAllAppTypes',
        'featuredApps',
        'loadingFeaturedApps',
      ]),
      ...mapState(externals_store, ['loadingExternalServices', 'externalServicesList']),
      ...mapState(ecommerce_store, ['loadingEcommerceApps', 'ecommerceAppsList']),
      searchOptions() {
        if (!this.allAppTypes || !this.externalServicesList) return [];

        const allApps = [...this.allAppTypes, ...this.externalServicesList];

        const filtered = allApps.filter((app) => {
          return app.name.toLowerCase().includes(this.searchTerm.trim().toLowerCase());
        });

        return filtered.map((app) => {
          return app.name;
        });
      },
      filteredApps() {
        if (!this.allAppTypes) return [];

        if (!this.searchTerm || !this.searchTerm.trim()) return this.allAppTypes;

        return this.allAppTypes.filter((app) => {
          return app.name.toLowerCase().includes(this.searchTerm.trim().toLowerCase());
        });
      },
      filteredExternalServices() {
        if (!this.externalServicesList) return [];

        if (!this.searchTerm || !this.searchTerm.trim()) return this.externalServicesList;

        return this.externalServicesList.filter((app) => {
          return app.name.toLowerCase().includes(this.searchTerm.trim().toLowerCase());
        });
      },
      filteredEcommerceApps() {
        if (!this.ecommerceAppsList) return [];

        if (!this.searchTerm || !this.searchTerm.trim()) return this.ecommerceAppsList;

        return this.ecommerceAppsList.filter((app) => {
          return app.name.toLowerCase().includes(this.searchTerm.trim().toLowerCase());
        });
      },
      hasAnyVisibleApp() {
        return (
          this.filteredExternalServices.length ||
          this.filteredEcommerceApps.length ||
          this.filteredApps.length
        );
      },
    },
    methods: {
      ...mapActions(app_type, ['getAllAppTypes', 'fetchFeatured']),
      ...mapActions(externals_store, ['getExternalServicesTypes']),
      ...mapActions(ecommerce_store, ['getEcommerceTypes']),
      async fetchChannels() {
        const params = {
          category: 'channel',
        };
        await this.getAllAppTypes({ params });

        if (this.errorAllAppTypes) {
          unnnic.unnnicCallAlert({
            props: {
              text: this.$t('apps.discovery.fetch_error'),
              type: 'error',
            },
            seconds: 6,
          });
          return;
        }

        this.channels.data = this.allAppTypes;
      },
      async fetchExternalServices() {
        await this.getExternalServicesTypes();
      },
      async fetchEcommerceApps() {
        await this.getEcommerceTypes();
      },
    },
  };
</script>

<style lang="scss" scoped>
  @import './styles.scss';
</style>
