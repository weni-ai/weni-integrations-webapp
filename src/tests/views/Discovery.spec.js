import { mount } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';
import Discovery from '../../views/Discovery/index.vue';
import { describe, expect, it, vi } from 'vitest';
import i18n from '@/utils/plugins/i18n';
import { defineStore } from 'pinia';

describe('Discovery', () => {
  it('matches snapshot', () => {
    const wrapper = mount(Discovery, {
      global: {
        plugins: [createTestingPinia(), i18n],
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('mounted', () => {
    it('should call fetchChannels', async () => {
      const spy = vi.spyOn(Discovery.methods, 'fetchChannels');
      const wrapper = mount(Discovery, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });
      await wrapper.vm.$nextTick();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call fetchExternalServices', async () => {
      const spy = vi.spyOn(Discovery.methods, 'fetchExternalServices');
      const wrapper = mount(Discovery, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });
      await wrapper.vm.$nextTick();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call fetchEcommerceApps', async () => {
      const spy = vi.spyOn(Discovery.methods, 'fetchEcommerceApps');
      const wrapper = mount(Discovery, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });
      await wrapper.vm.$nextTick();
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should call fetchFeatured', async () => {
      const spy = vi.spyOn(Discovery.methods, 'fetchFeatured');
      const wrapper = mount(Discovery, {
        global: {
          plugins: [createTestingPinia({ createSpy: vi.fn }), i18n],
        },
      });
      await wrapper.vm.$nextTick();
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('fetchChannels', () => {
    it('should call getAllTypes', async () => {
      const pinia = createTestingPinia({
        stubActions: false,
        createSpy: vi.fn,
      });

      const useAppTypeStore = defineStore('appType', {
        state: () => ({
          allAppTypes: [{ name: 'Test App' }],
          errorAllAppTypes: false,
        }),
        actions: {
          async getAllAppTypes() {
            return Promise.resolve();
          },
        },
      });

      const wrapper = mount(Discovery, {
        global: {
          plugins: [pinia, i18n],
        },
      });
      const appTypeStore = useAppTypeStore();

      await wrapper.vm.fetchChannels();
      expect(appTypeStore.getAllAppTypes).toHaveBeenCalledTimes(2);
    });
  });
});
