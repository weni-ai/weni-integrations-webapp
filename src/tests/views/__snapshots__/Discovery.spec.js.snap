// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Discovery > matches snapshot 1`] = `
"<div data-v-223130b3="" class="discovery-content">
  <unnnic-input data-v-223130b3="" modelvalue="" class="discovery-content__search" placeholder="Search" icon-left="search-1"></unnnic-input>
  <!--v-if-->
  <div data-v-223130b3="" class="discovery-content__grids">
    <div data-v-deb56cf9="" data-v-223130b3="">
      <section data-v-deb56cf9="" id="app-grid" maxlength="255" disabled="false">
        <!--v-if-->
        <div data-v-deb56cf9="" class="app-grid__content"></div>
        <!--v-if-->
      </section>
      <unnnic-modal data-v-deb56cf9="" showmodal="false" text="Remove Application" scheme="feedback-red" modal-icon="alert-circle-1"></unnnic-modal>
      <div data-v-7d3d5e15="" data-v-deb56cf9="">
        <transition-stub data-v-7d3d5e15="" name="fade" appear="false" persisted="false" css="true">
          <!--v-if-->
        </transition-stub>
        <unnnic-modal data-v-7d3d5e15="" showmodal="false" text="Cancel modifications" scheme="feedback-yellow" modal-icon="alert-circle-1"></unnnic-modal>
      </div>
    </div>
    <div data-v-deb56cf9="" data-v-223130b3="">
      <div data-v-ab3bf254="" data-v-deb56cf9="">
        <div data-v-ab3bf254="" class="title">
          <unnnic-skeleton-loading data-v-ab3bf254="" tag="div" width="40px" height="40px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" tag="div" width="208px" height="40px"></unnnic-skeleton-loading>
        </div>
        <div data-v-ab3bf254="" class="grid">
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
        </div>
      </div>
      <unnnic-modal data-v-deb56cf9="" showmodal="false" text="Remove Application" scheme="feedback-red" modal-icon="alert-circle-1"></unnnic-modal>
      <div data-v-7d3d5e15="" data-v-deb56cf9="">
        <transition-stub data-v-7d3d5e15="" name="fade" appear="false" persisted="false" css="true">
          <!--v-if-->
        </transition-stub>
        <unnnic-modal data-v-7d3d5e15="" showmodal="false" text="Cancel modifications" scheme="feedback-yellow" modal-icon="alert-circle-1"></unnnic-modal>
      </div>
    </div>
    <div data-v-deb56cf9="" data-v-223130b3="">
      <div data-v-ab3bf254="" data-v-deb56cf9="">
        <div data-v-ab3bf254="" class="title">
          <unnnic-skeleton-loading data-v-ab3bf254="" tag="div" width="40px" height="40px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" tag="div" width="208px" height="40px"></unnnic-skeleton-loading>
        </div>
        <div data-v-ab3bf254="" class="grid">
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
          <unnnic-skeleton-loading data-v-ab3bf254="" class="grid__item" tag="div" width="256px" height="200px"></unnnic-skeleton-loading>
        </div>
      </div>
      <unnnic-modal data-v-deb56cf9="" showmodal="false" text="Remove Application" scheme="feedback-red" modal-icon="alert-circle-1"></unnnic-modal>
      <div data-v-7d3d5e15="" data-v-deb56cf9="">
        <transition-stub data-v-7d3d5e15="" name="fade" appear="false" persisted="false" css="true">
          <!--v-if-->
        </transition-stub>
        <unnnic-modal data-v-7d3d5e15="" showmodal="false" text="Cancel modifications" scheme="feedback-yellow" modal-icon="alert-circle-1"></unnnic-modal>
      </div>
    </div>
    <div data-v-deb56cf9="" data-v-223130b3="">
      <section data-v-deb56cf9="" id="app-grid" maxlength="255" disabled="false">
        <div data-v-deb56cf9="" class="app-grid__header">
          <unnnic-avatar-icon data-v-deb56cf9="" icon="bar_chart_4_bars" scheme="aux-red-500" filled="true" opacity="false" size="sm" class="app-grid__header__icon"></unnnic-avatar-icon>
          <p data-v-deb56cf9="" class="app-grid__header__title">BI Tools</p>
        </div>
        <div data-v-deb56cf9="" class="app-grid__content">
          <unnnic-card data-v-deb56cf9="" class="app-grid__content__item" type="marketplace" title="Power BI" description="Keep up with the growth of your project with the Power BI connector with the Weni Platform." comments="undefined comments" rating="0" iconsrc="/src/assets/logos/power_bi.png" typeaction="edit" clickable="true"></unnnic-card>
        </div>
        <!--v-if-->
      </section>
      <unnnic-modal data-v-deb56cf9="" showmodal="false" text="Remove Application" scheme="feedback-red" modal-icon="alert-circle-1"></unnnic-modal>
      <div data-v-7d3d5e15="" data-v-deb56cf9="">
        <transition-stub data-v-7d3d5e15="" name="fade" appear="false" persisted="false" css="true">
          <!--v-if-->
        </transition-stub>
        <unnnic-modal data-v-7d3d5e15="" showmodal="false" text="Cancel modifications" scheme="feedback-yellow" modal-icon="alert-circle-1"></unnnic-modal>
      </div>
    </div>
  </div>
  <!--v-if-->
</div>"
`;
